tests/test_function_utils.py

```diff
@@ -9,35 +12,68 @@
-        def returns_1(param_1, param_2, param_3=11, param_4=22):
+        def function_with_params(param_1, param_2, param_3=11, param_4=22):
             return 1
 
-        key1 = build_value_by_func_parameters('asdfghjklzxcvbnm', returns_1, *[1234, 12345134], **{})
+        key1 = build_value_by_func_parameters('asdfghjklzxcvbnm', function_with_params, *[1234, 12345134], **{})
         self.assertNotEqual(key1, None)
         self.assertEqual(key1, 'asdfghjklzxcvbnm')
 
-        key2 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', returns_1, *[11, 22], **{})
+        key2 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', function_with_params, *[11, 22], **{})
         self.assertNotEqual(key2, None)
         self.assertEqual(key2, 'xyz_11_22')
 
-        key3 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', returns_1, *[11], **{})
+        key3 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', function_with_params, *[11], **{})
         self.assertNotEqual(key3, None)
         self.assertEqual(key3, 'xyz_11_!Eparam_2E!')
 
-        key4 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', returns_1, *[], **{'param_2': 'pp2'})
+        key4 = build_value_by_func_parameters(
+            'xyz_{param_1}_{param_2}', function_with_params, *[], **{'param_2': 'pp2'}
+        )
         self.assertNotEqual(key4, None)
         self.assertEqual(key4, 'xyz_!Eparam_1E!_pp2')
 
-        key5 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', returns_1, 1, 2)
+        key5 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', function_with_params, 1, 2)
         self.assertNotEqual(key5, None)
         self.assertEqual(key5, 'xyz_1_2')
 
-        key6 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', returns_1, None, None)
+        key6 = build_value_by_func_parameters('xyz_{param_1}_{param_2}', function_with_params, None, None)
         self.assertNotEqual(key6, None)
         self.assertEqual(key6, 'xyz_!Eparam_1E!_!Eparam_2E!')
 
+        dat = datetime.date(2022, 1, 1)
+        dattime = datetime.datetime(2022, 1, 1)
+        tpl = ("fruit", "apple")
+        lst = ["fruit", "apple"]
+        point = WKTElement('POINT(5 45)')
+        set_value = {"fruit", "apple"}
+        obj = {"fruit1": "apple", "fruit2": "orange"}
+
+        def function_with_multi_params(
+            param_1, param_2, param_3, param_4, param_5, param_6=point, param_7=set_value, param_8=MyEnum.one.value
+        ):
+            pass
+
+        key7 = build_value_by_func_parameters(
+            'xyz_{param_1}_{param_2}_{param_3}_{param_4}_{param_5}_{param_6}_{param_7}_{param_8}',
+            function_with_multi_params,
+            dat,
+            dattime,
+            tpl,
+            lst,
+            obj,
+        )
+        self.assertNotEqual(key7, None)
+        self.assertEqual(key7, f'xyz_{dat}_{dattime}_{tpl}_{lst}_{obj}_{point}_{set_value}_{MyEnum.one.value}')
```

Comment:
1. rename `build_value_by_func_parameters` to `build_function_key`
2. use to return the value of the object 
3. use `repr` to get a string representation of the value.   (UnseenFormatter)
4. create new abstract class called `Keyable` inside the `func_utils` file (UnseenFormatter)
```
class Keyable:
  def key(self) -> str:
     pass
```
5. if a value is instance of `Keyable` then get the value from the `key` function
6. add this exmaple to the docs 
```
class User(Keyable):

  def __init__(self, id, name):
     self.id = ...
     self.name = ...

  def key(self):
     return f"User(name={[self.name](http://self.name/)},id={[self.id](http://self.id/)}"
```